guide/popular-topics/errors.md

```diff
@@ -0,0 +1,155 @@
+# Errors
+
+There is no doubt that you have encountered errors while making bots. While errors are very useful at warning you of what is going wrong, many people are stumped by them and how to track them down and fix them, but don't worry, we have you covered. This section will be all about how to diagnose errors, how to identify where they are coming from, and how to fix them.
+
+## Types of Errors
+
+### API Errors
+
+API Errors or DiscordAPIErrors are errors which are thrown by the Discord API when an invalid request is carried out. API Errors can be mostly diagnosed using the message that is given. They can also be further examined by seeing the http method and path used, we will explore tracking these errors down in the next section.
+
+Example: `DiscordAPIError: Cannot send an empty message`
+
+### Discord.js Errors
+
+Discord.js Errors are errors which are thrown by the library itself, they can usually be easily tracked down using the stacktrace and error message.
+
+Example: `Error: Members didn't arrive in time.`
+
+### JS Errors
+
+JS Errors are simple errors which can be thrown by node itself, or by discord.js. These types of errors can easily be fixed by looking at the type of error, and the stacktrace. You can find a full list of types [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) And a list of common js errors [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors).
+
+Example: `ReferenceError: "x" is not defined`
+
+### Websocket and Network Errors
+
+Websocket and Network errors are common system errors thrown by Node in response to something wrong with the websocket connection. Unfortunately, these errors do not have a concrete solution and can be (usually) fixed by getting a better, and stronger connection. However, if you handle these errors correctly, discord.js will attempt to reconnect to the websocket automatically.
```

Comment:
This isn't really accurate, definitely not on master. We are handling network errors internally and attempt to reconnect if you have a network blip, in order to reconnect you the moment we can. 

> Unfortunately, these errors do not have a concrete solution and can be (usually) fixed by getting a better, and stronger connection

More-less; you need a stable connection too!

> However, if you handle these errors correctly, discord.js will attempt to reconnect to the websocket automatically.

In master, we emit `shardError` for any WS error we receive from the WS, the rest being handled internally to ensure you, the end user, have no issues even if your internet randomly dies. I don't know how applicable this is to stable d.js so don't quote me

Reply:
i know there are a lot of websocket changes between stable and master, so i was not very sure if master already handled reconnecting without the error listener. For master should i mention you should listen to `shardError` to be notified about websocket errors? and that the lib will already attempt to reconnect?

Reply:
Both versions handle reconnects automatically, the issue is that stable emits the `error` event, which by default throws if not handled (thus, crashing). 

You don't *have* to listen to `shardError` at all! The lib should work regardless of it

Feel free to amend it like the the MessageField changes if you're reading the v12.x branch, if you want

Reply:
yeah i think ill refactor master websocket stuff to say 
`Master will automatically handle websocket connection errors and attempt to reconnect, if you would like to do something based on these errors, you should listen to shardError`, does that look good to you?

Reply:
``On master, WebSocket errors are handled internally, meaning your process should never crash from them. If you want to log these errors, should they happen, you can listen to the `shardError` event``

Reply:
sounds good